### PR TITLE
feat: Add support for count(*) aggregation query firestore

### PR DESCRIPTION
**Pull request by hemanshv**
Adding count(*) functionality in firestore

### CODE REVIEW

**Comment by hemanshv** [comment]
Hi @jskeet @amanda-tarafa ,
Its a draft PR for query count feature. Its still under process. 
I have raised the PR to get some internal feedback on the same.
Please do not review. Will let you know once the changes are finalized.
Thanks,

**Comment by Rishabh-V** [comment]
I will have a look later today after going through the design document. Thanks.

**Review by Rishabh-V — COMMENTED** [review]
First pass done. I have only reviewed the production code for now based on my limited understanding after reading the design doc. Will sync up with you early next week when you are back and we can do a second pass for the complete codebase after clarifying my understanding. Thanks.

**Review by Rishabh-V — COMMENTED** [review]
I have reviewed based on my understanding. Once you address/respond to these comments, please feel free to remove the LEX-review label and request Jon and Amanda for their review. Thank you.

**Comment by hemanshv** [comment]
Hi @jskeet  @amanda-tarafa ,
This is now ready for review. 
Please review and provide you feedback.

Thanks,

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:114]: No need for this - the type is sealed.

Again though, I'd encourage you to include the aggregations as part of the state, both here and for GetHashCode.
If we represent "we've asked for count" as state, even though it's the only available aggregation available now, it'll make it much easier to add support for other aggregations later.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuerySnapshot.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuerySnapshot.cs:36]: Possibly add a note about limits?
What will this return in the future when we support non-count aggregations, and a count hasn't been specified?
(Perhaps the return type should actually be `long?` instead?) I'm genuinely not sure.

**Review by jskeet — COMMENTED** [review]
Main comments are around the way that AggregateQuery is tested, and the current lack of state indicating the kind of aggregation being used.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuerySnapshot.cs
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuerySnapshot.cs:36]: Thanks for your feedback @jskeet .Will update it and let you know.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/QueryTest.cs
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/QueryTest.cs:447]: No there is nothing in the result which indicate the same, but I think its trivial. If the limit provided is equal to the count then we can say that there is "at least n number of records".

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/QueryTest.cs
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/QueryTest.cs:446]: No not really.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/QueryTest.cs
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/QueryTest.cs:447]: The problem is that the code that supplies the limit may not be the code that handles the response. I don't think we need any change in this PR, but it *might* be worth asking the Firestore team to consider indicating somewhere in the response that the count is not necessarily accurate.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:114]: Make Sense Jon will do.
Thanks

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuerySnapshot.cs
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuerySnapshot.cs:36]: Yes Jon I will add a note about the limit thing. On return thing: When we support non-count aggregations and count hasn't been specified in that case the value of count would be "null". I have updated the return type to be long?.
Thanks

**Comment by jskeet** [comment]
Will look when I get the chance - even coming back from one day off, I've got a bit of a backlog...

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:54]: Rather than splitting the initialization of the Count aggregation, why not create an internal `CreateCountAggregation` method in here, which can create it fully?

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:58]: Rather than allowing _aggregations to *ever* be null, I'd initialize it with an empty list (which you only need one instance of) in the single-parameter constructor. That should simplify all code using it.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:82]: You can simplify this to

```csharp
readTime ??= Timestamp.FromProtoOrNull(response.ReadTime);
```

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:78]: I'm not sure this is what you want - that means if there are multiple responses with `AggregateFields` populated, if the count is in the first one, we'll end up losing the count. You could do something like:

```csharp
if (count is null)
{
    response.Result.AggregateFields?.TryGetValue(CountAlias, out count);
}
```

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:78]: Actually, I'd also keep `count` as a `long?` variable, and have a check here - so something like:

```csharp
if (count is null && response.Result.AggregateFields?.TryGetValue(CountAlias, out var countValue) == true)
{
    // You probably need to change this - I can't remember exactly what it would be, but hopefully it'll be obvious enough in an IDE.
    GaxPreconditions.CheckState(countValue.ValueCase == Value.Types.ValueCase.Integer, "The count was not an integer.");
    count = countValue.IntegerValue;
}
```

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:100]: You could just use `Transaction = transactionId ?? ByteString.Empty` in the object initializer, I suspect.
(And yes, this may be something that could be simplified in Query too... I haven't checked.)

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:76]: It might be worth extracting this lambda expression into a local method (e.g. `ProcessResponse`) for readability - I probably didn't do it when writing the Firestore client initially as they didn't exist back then ;)

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuerySnapshot.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuerySnapshot.cs:36]: The second and third sentences should be a single sentence, something like:

"Note that when the query contains a limit, the count is restricted by that limit."

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuerySnapshot.cs:36]: I'd also document the conditions in which it will be null (anticipating that actually being feasible...)

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Query.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Query.cs:427]: Right, and this would become `new AggregateQuery(this).WithAggregation(AggregateQuery.CreateCountAggregate))`. Or you might introduce an `Aggregates` static class as a factory for aggregates. (You'd need to expose the count alias in there, of course.)

**Comment by hemanshv** [comment]
Thanks for the feedback Jon. I will address them and let you know.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/TransactionTest.cs
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/TransactionTest.cs:108]: I missed the validation of writes in our new method inside transaction. This should fail as per our current transaction semantics(all reads to be executed before writes). I will update that. Thanks for pointing this out.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/AggregateQueryTest.cs
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/AggregateQueryTest.cs:66]: Could you please elaborate a bit more on this.

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/AggregateQueryTest.cs:66]: Which part? The first is just "use some separate local variables for clarity instead of doing everything in a single huge statement".
The second is about creating entirely independent but equal queries: `s_db.Collection("xyz")` and calling `s_db.Collection("xyz")` again, so equal underlying queries rather than a single Query object used in both cases.

Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/AggregateQueryTest.cs:66]: Got it Thanks

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:100]: While using this we are encountering StatusCode="InvalidArgument", Detail="Invalid transaction." error. The error is occurring because the setter of transactionId also sets consistencySelectorCase_  to ConsistencySelectorOneofCase.Transaction and a transaction with an empty transaction id is an invalid transaction. So I think we should stick to the old way.

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:100]: Ah, right. Yup, that makes sense.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/AggregateQuerySnapshotTest.cs
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/AggregateQuerySnapshotTest.cs:37]: Much better now, thanks!

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/AggregateQuerySnapshotTest.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/AggregateQuerySnapshotTest.cs:51]: Super-nit: Please start comments with a capital letter, even if they're just sentence fragments.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/AggregateQuerySnapshotTest.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/AggregateQuerySnapshotTest.cs:55]: Super nit: "read time" instead of "readtime" (it's logically two words, which is why the T is capitalized in the PascalCasedName).

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:90]: Nit: space before `(`

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:68]: Style nit we can discuss in our meeting - I personally tend to put local methods at the end of the method (with a line break before each one). I think that allows the code to stay in blocks of the same abstraction level. Let's talk about it though.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs:79]: Nit: Maybe just return directly rather than using another local variable here?

**Review by jskeet — APPROVED** [review]
Just tiny nits now - I've approved so that you can merge after fixing those, or we could even address them in a separate PR if you want. (Please squash to a single commit before merging though.)

**Pull request merged**
